src/plugin.ts

```diff
@@ -57,8 +57,32 @@ export class Plugin {
     this.pluginPath = path;
 
     this.version = packageJSON.version || "0.0.0";
-    this.main = packageJSON.main || "./index.js"; // figure out the main module - index.js unless otherwise specified
     this.isESM = packageJSON.type === "module";
+    this.main = "";
+
+    // figure out the main module
+    // exports is available - https://nodejs.org/dist/latest-v14.x/docs/api/packages.html#packages_package_entry_points
+    if (packageJSON.exports) {
+      // main entrypoint - https://nodejs.org/dist/latest-v14.x/docs/api/packages.html#packages_main_entry_point_export
+      if (typeof packageJSON.exports === "string") {
+        this.main = packageJSON.exports;  
+      } else { // subpath export - https://nodejs.org/dist/latest-v14.x/docs/api/packages.html#packages_subpath_exports
+        // conditional exports - https://nodejs.org/dist/latest-v14.x/docs/api/packages.html#packages_conditional_exports
+        const exports = packageJSON.exports.import || packageJSON.exports.require || packageJSON.exports.node || packageJSON.exports.default || packageJSON.exports["."];
```

Comment:
Also applying to the line below which handles condition exports.

I'm not quite sure if this simple handling for conditional exports will work?
The `exports.import` is used to supply the entry point for ESM modules, right? And `exports.require` is used to supply the "regular" require location.
Meaning, depending on the value we use, we would need to adjust the `isESM` property, do we?

While I'm not sure what the right decision here is, we should definitely create some documentation around `exports` field on https://developers.homebridge.io/ (and in the according template repos) when writing homebridge plugins.

E.g. https://nodejs.org/dist/latest-v14.x/docs/api/packages.html#packages_package_entry_points writes:
`
Warning: Introducing the "exports" field prevents consumers of a package from using any entry points that are not defined, including the package.json (e.g. require('your-package/package.json'). This will likely be a breaking change.
`
This is required for homebridge plugins.
Can we detect such an error, when `package.json` wasn't included in the `exports` declaration, such that we can log an error explaining the situation?

Reply:
I'm not sure either on how to treat the package when both `exports.import` and `exports.require` are defined. I assumed that if a package uses both, then it works in both ESM and CJS. Since we support both, loading one of them does not make any difference and thus I stopped at first found.

About the warning about being not able to require "private" files, I don't think it specifically applies to homebridge. First of all, we load `package.json` via `fs.readFileSync`, for which this restriction does not apply (that one applies only to `require`), then we only load the main entrypoint of a package.

For package developers that try to include private files will receive an error from node, but I don't think we should have a doc for all possible error coming from using Node, isn't it? :)

Reply:
@Supereg @oznu Are we good on this? Can we land this PR?

Reply:
I'm not sure if my point was addressed yet?

> The exports.import is used to supply the entry point for ESM modules, right? And exports.require is used to supply the "regular" require location.
Meaning, depending on the value we use, we would need to adjust the isESM property, do we?


> "I'm not sure either on how to treat the package when both exports.import and exports.require are defined."

Not sure if you may misunderstood my point, I was actually not talking about the situation where both are defined, but e.g. when only `exports.import` is defined (implying a ESM module). Would this require us to set the `isESM` property (which currently is only set depending on the package.json `type` field)? Or is the `type` property set to `module` anyways?

Apart from that point, where I'm unsure myself on how to handle that situation (haven't yet encountered any packages using the `exports` property), I'm happy to land this PR.
Maybe @oznu has some expertise on this?

Reply:
Oh, I see your point now. I think I misunderstood it.
No, `package.exports` is also available for CommonJS. Only `type` explicitly sets the type to be ESM.

Reply:
> No, package.exports is also available for CommonJS.

Yes that's true. However `exports.import` will point to a ESM module file. As of right now, we require some workaround to import ESM modules (see #2915 and https://github.com/microsoft/TypeScript/issues/43329).

To my understanding this needs to be handled. I may be wrong though.
Could you verify that this does indeed work? Or have you already tested it thoroughly?

Reply:
I forgot to test te `exports.import` case. If it is set now I force `isESM` otherwise it will not work below.
The `await import` of the TS file works fine instead.

Reply:
@Supereg All clear now? Can I rebase so we finally land this?